service/controller/release/resource/argoapps/resource.go

```diff
@@ -0,0 +1,255 @@
+		if appsToCreate.Items[i].Labels == nil {
+			appsToCreate.Items[i].Labels = map[string]string{}
+		}
+		appsToCreate.Items[i].Labels[key.LabelManagedBy] = project.Name()
+
+		err := r.k8sClient.CtrlClient().Create(
+			ctx,
+			&appsToCreate.Items[i],
+		)
+		if apierrors.IsAlreadyExists(err) {
+			// fall through.
+		} else if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("created app %#q in namespace %#q", app.GetName(), app.GetNamespace()))
+	}
+
+	return nil
+}
+
+func calculateMissingApps(componentApps map[string]argoappv1alpha1.Application, apps argoappv1alpha1.ApplicationList) argoappv1alpha1.ApplicationList {
+	var missingApps argoappv1alpha1.ApplicationList
+
+	for _, componentApp := range componentApps {
+		missingApp := true
+		for _, app := range apps.Items {
+			if compareArgoApplications(componentApp, app) {
+				missingApp = false
+				break
+			}
+		}
+		if missingApp {
+			missingApps.Items = append(missingApps.Items, componentApp)
+		}
+	}
+
+	return missingApps
+}
+
+func calculateObsoleteApps(componentApps map[string]argoappv1alpha1.Application, apps argoappv1alpha1.ApplicationList) argoappv1alpha1.ApplicationList {
+	var obsoleteApps argoappv1alpha1.ApplicationList
+
+	for _, app := range apps.Items {
+		obsoleteApp := true
+		for _, componentApp := range componentApps {
+			if compareArgoApplications(componentApp, app) {
+				obsoleteApp = false
+				break
+			}
+		}
+		if obsoleteApp {
+			obsoleteApps.Items = append(obsoleteApps.Items, app)
+		}
+	}
+
+	return obsoleteApps
+}
+
+func compareArgoApplications(a, b argoappv1alpha1.Application) bool {
+	if a.Name != b.Name {
+		return false
+	}
+
+	konfigureA := extractKonfigureVariables(a)
+	konfigureB := extractKonfigureVariables(b)
+
+	match := true
+	match = match && konfigureA.AppName != "" && konfigureA.AppName == konfigureB.AppName
+	match = match && konfigureA.AppVersion != "" && konfigureA.AppVersion == konfigureB.AppVersion
+	match = match && konfigureA.AppCatalog != "" && konfigureA.AppCatalog == konfigureB.AppCatalog
+
+	return match
+
+}
+
+type konfigureVariables struct {
+	AppName    string
+	AppVersion string
+	AppCatalog string
+}
+
+func extractKonfigureVariables(app argoappv1alpha1.Application) *konfigureVariables {
+	v := &konfigureVariables{}
+
+	if app.Spec.Source.Plugin.Name != "konfigure" {
+		return v
+	}
+
+	for _, envVar := range app.Spec.Source.Plugin.Env {
+		switch envVar.Name {
+		case "KONFIGURE_APP_NAME":
+			v.AppName = envVar.Value
+		case "KONFIGURE_APP_VERSION":
+			v.AppVersion = envVar.Value
+		case "KONFIGURE_APP_CATALOG":
+			v.AppCatalog = envVar.Value
+		}
+	}
+
+	return v
+}
+
+func (r *Resource) componentToArgoApplication(ctx context.Context, component releasev1alpha1.ReleaseSpecComponent) (app *argoappv1alpha1.Application, err error) {
+	configRef := "main"
+	if component.ConfigReference != "" {
+		configRef = component.ConfigReference
+	}
```

Comment:
Don't you trust the defaulting here: https://github.com/giantswarm/apiextensions/pull/792/files#diff-ce130929b840710ed312dfe6b4eccc25af1e3f748bbda4434caa2a9c6baed2efR102 ?

Reply:
:detective: Defaulting is fine. This is in case `ConfigReference` field is deleted entirely.